### ISSUE TITLE
Lack of Undo/Redo functionality

### ISSUE THREAD

**Issue by OverloadedOrama** [OverloadedOrama/Pixelorama#8]
I tried implementing an undo/redo system which worked, until I implemented the layer system. I gave up on it ever since so I could focus on adding new features, but it is a vital feature to have. And now that there are so many features compared to the past, it's going to be even harder. Any help would be appreciated, whether it's just ideas of what to do or even pull requests.

**Comment by OverloadedOrama** [comment]
Started working on it as of 7b8c6bbf000752cafa33135f70a6365e7f282a05. I use Godot's UndoRedo class so it can be (theoretically) more easily ported as a plug-in in the future. Check the commit description for more details.

**Comment by OverloadedOrama** [comment]
Closing this issue as Pixelorama now has a functioning UndoRedo system as of 40e0978b3f2fffa17910dcdb31d1001e37dec055. It could use some improvements, such as maybe it could take into account actions such as layer/frame adding/removing/cloning/moving/etc, button presses and all that. Right now it mostly deals with drawing itself, and edit menu tools like cropping, scaling, flipping. Feel free to open another issue or do pull requests if you'd like to see improvements.

I also hope it'll work if Pixelorama is ever to be ported as a Godot plug-in. If you encounter bugs, you can open a new issue or do pull requests.

**Comment by OverloadedOrama** [comment]
Re-opening this because I realized that if layer and frame adding/removing/cloning/moving/etc don't have UndoRedo, it will lead to weird results. Now, I have successfully managed to implement this functionality for layers (ec7b2129ff5277b9a94270629470370b567d17d2), but frames are harder, because it deals with creating new canvas instances and them freeing them from memory. Since it is impossible to access a node that has been deleted, I can't just "undo" a frame removal.

I have some ideas for workarounds, and hopefully I'll commit them in the next few days. Otherwise if anyone is adventurous enough to give it a shot, go ahead!
